### PR TITLE
3d-tiles

### PR DESCRIPTION
**Pull request by Jeremy-Gaillard**
Early support for 3d-tiles. To try the demo (examples/3dtiles.html) you need to run the 3d-tiles sample server provided [here](https://github.com/AnalyticalGraphicsInc/3d-tiles-samples).

PR for visibility and feedback, I'm still working on a few bugs.

What coverage of the 3d-tiles spec should we have before merging? See #185 for progress.

### CODE REVIEW

File: examples/3dtiles.html [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                    return [];
+                }
+                itowns.pre3DTilesUpdate(context, layer);
+                return [layer.root];
+            };
+            threeDTilesLayer.update = itowns.process3DTilesNode(
+                itowns.threeDTilesCulling,
+                itowns.threeDTilesSubdivisionControl
+            );
+            threeDTilesLayer.url = 'http://localhost:8003/tilesets/TilesetWithDiscreteLOD/tileset.json';
```
Comment (peppsac) [inline examples/3dtiles.html]: Where can I get test files like this one?

Comment (Jeremy-Gaillard) [reply examples/3dtiles.html]: The test files are those provided by AGI in their [3d-tiles-samples repository](https://github.com/AnalyticalGraphicsInc/3d-tiles-samples). Just run the provided node server locally and it should work.

Comment (qdnguyen) [reply examples/3dtiles.html]: i dit it, but any b3dm is loaded, how can we see the dragon model? we must move to a specific place?

Comment (Jeremy-Gaillard) [reply examples/3dtiles.html]: The camera is a bit misplaced, you should see the model if you move slightly to the top-right of the screen. I will fix this ASAP.

**Comment by Jeremy-Gaillard** [comment]
Fixed position of the camera and bad mesh orientation.

**Comment by Jeremy-Gaillard** [comment]
I'd like to remove glTFLoader from the code base and add it as a dependency. The loader is currently in the examples folder of three (https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/GLTFLoader.js). Any idea how I could do this? (ping @autra)

**Comment by peppsac** [comment]
Hmm visibility testing is broken for the globe :s
(the problem was there before the changes from 6e5ae5bf689a5564fe22e07a12c9f27a90059bbe )

We definitely need to fix this...

**Comment by peppsac** [comment]
Globe occlusion issue fixed in PR https://github.com/iTowns/itowns2/pull/307

File: src/Core/Scheduler/Providers/ThreeDTiles_Provider.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+ */
+import * as THREE from 'three';
+import Provider from './Provider';
+import CacheRessource from './CacheRessource';
+import B3dmLoader from '../../../Renderer/ThreeExtended/B3dmLoader';
+import Fetcher from './Fetcher';
+import FeatureMaterial from '../../../Renderer/FeatureMaterial';
+
+
+export function ThreeDTilesIndex(tileset) {
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/ThreeDTiles_Provider.js]: I think I'd prefer `_3dTilesXXXX` names over, `ThreeDTilesXXX`.

Cons:
   * `_` is supposed to be reserved for private usage
Pros:
  * names that match the procol/spec name
  * _3dTilesIndex is more readable than ThreeDTilesIndex

WDYT?

File: src/Renderer/Shader/FeatureVS.glsl [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+uniform int        RTC;
+varying float      light;
+
+attribute vec3 color;
+varying vec3 vColor;
+
+// IE error : Initializer for const variable must initialize to a constant value
+//const vec3 dir =  normalize(vec3(1.0,1.0,0.5));
+
+void main()
```
Comment (peppsac) [inline src/Renderer/Shader/FeatureVS.glsl]: I fail to see why makes this file needed?

Comment (Jeremy-Gaillard) [reply src/Renderer/Shader/FeatureVS.glsl]: It's to transfer color information to the fragment shader.

I think I'll remove feature material and shaders from this PR before it's merged, they aren't really useful or relevant to this PR.

File: src/Core/Scheduler/Providers/ThreeDTiles_Provider.js (hunk 2 above)
Comment (Jeremy-Gaillard) [reply src/Core/Scheduler/Providers/ThreeDTiles_Provider.js]: I agree with ThreeDTiles being an awkward name. I can't think of better names than yours, so I'll use them next time I refactor.

File: examples/3dtiles.html (hunk 1 above)
Comment (qdnguyen) [reply examples/3dtiles.html]: no thing shows when i move the camera, no connection between itowns and 3d-tiles sample server, have any ideas?

Comment (Jeremy-Gaillard) [reply examples/3dtiles.html]: Is your branch up-to-date? Did you open the right demo (localhost:8080/examples/3dtiles.html)?

**Comment by autra** [comment]
> I'd like to remove glTFLoader from the code base and add it as a dependency. The loader is currently in the examples folder of three (https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/GLTFLoader.js). Any idea how I could do this? (ping @autra)

You can import things with a path starting with the package name. `import GltfLoader from three/examples/js/loaders/GLTFLoader` should work (I haven't tested).

File: examples/3dtiles.html (hunk 1 above)
Comment (qdnguyen) [reply examples/3dtiles.html]: ok, thanks! it work now, i can see dragon model! but i found some bugs, the globe disappear when we zoom in and the left mouse button is bloked. For second bug, if we use right mouse button to move the camera, after that, the left mouse come back. Anyway, good job, i will test with other models and give you more feedbacks.

File: src/Core/Scheduler/Providers/ThreeDTiles_Provider.js (hunk 2 above)
Comment (autra) [reply src/Core/Scheduler/Providers/ThreeDTiles_Provider.js]: AFAIK variables in javascript can start with most unicode character. We *could* find a char that looks like a 3 in other language, it would be funny for everyone involved ;-)

Jokes apart, why not following jquery example and start with `$`? `$3DTiles` etc... It is somehow familiar for a js dev, and has the advantage of not looking like a private var.

Comment (autra) [reply src/Core/Scheduler/Providers/ThreeDTiles_Provider.js]: Other possibility, roman numerals: `iiiDTiles` (but I think I still prefer the `$`-prefixed version)

**Comment by Jeremy-Gaillard** [comment]
@autra thanks for the answer. Unfortunately, it doesn't work perfectly.

I tried to do this:
```javascript 
import GlTFLoader from 'three/examples/js/loaders/GLTFLoader';
``` 
However, since the examples don't use ES6 imports, it only sets the value of THREE.GLTFLoader, and since THREE isn't defined in my class, it fails.

The following code works:
```javascript 
import * as THREE from 'three';
import GlTFLoader from 'three/examples/js/loaders/GLTFLoader';
...
this.glTFLoader = new THREE.GLTFLoader();
``` 
But I have a jslint error since GlTFLoader isn't used in my code. I guess I could deactivate this specific warning, but it seems a bit hacky. Do you have a better idea?

File: src/Core/Scheduler/Providers/ThreeDTiles_Provider.js (hunk 2 above)
Comment (Jeremy-Gaillard) [reply src/Core/Scheduler/Providers/ThreeDTiles_Provider.js]: $3dTiles suits me. I'll make the change.

**Comment by Jeremy-Gaillard** [comment]
Updated.

I think the branch is in a stable enough state to be considered for merging now.

**Comment by autra** [comment]
> The following code works:
> 
> ```js
> import * as THREE from 'three';
> import GlTFLoader from 'three/examples/js/loaders/GLTFLoader';
> ...
> this.glTFLoader = new THREE.GLTFLoader();
> ```
> But I have a jslint error since GlTFLoader isn't used in my code. I guess I could deactivate this specific warning, but it seems a bit hacky. Do you have a better idea?

Yep, it works because of the `imports` loader in webpack.config.js. Also, `import  'three/examples/js/loaders/GLTFLoader'` [should work too](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) and pass eslint.

**Comment by Jeremy-Gaillard** [comment]
Indeed it does. Fixed it.

File: src/Core/Scheduler/Providers/$3dTiles_Provider.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+ * Description:
+ */
+import * as THREE from 'three';
+import Provider from './Provider';
+import B3dmLoader from '../../../Renderer/ThreeExtended/B3dmLoader';
+import Fetcher from './Fetcher';
+import BasicMaterial from '../../../Renderer/BasicMaterial';
+
+
+export function $3dTilesIndex(tileset, urlPrefix) {
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/$3dTiles_Provider.js]: The name `$3dTilesIndex` is fine but I'd suggest renaming the files and remove their `$` prefix.

File: src/Core/Scheduler/Providers/$3dTiles_Provider.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
+/**
+ * Created On: 2016-11-8
+ * Class: $3dTiles_Provider
+ * Description:
+ */
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/$3dTiles_Provider.js]: Please remove this

File: src/Core/Scheduler/Providers/$3dTiles_Provider.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        const e = center.x + box[3];
+        const s = center.y - box[7];
+        const n = center.y + box[7];
+        const b = center.z - box[11];
+        const t = center.z + box[11];
+
+        return { box: new THREE.Box3(new THREE.Vector3(w, s, b), new THREE.Vector3(e, n, t)) };
+    } else if (boundingVolume.sphere) {
+        // TODO
+        return { sphere: undefined };
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/$3dTiles_Provider.js]: Maybe all the unimplemented code path should `throw new Error()..)` instead of silently failing?

File: src/Core/Scheduler/Providers/$3dTiles_Provider.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+function getBox(boundingVolume) {
+    if (boundingVolume.region) {
+        // TODO: unsupported
+        return { region: boundingVolume.region };
+    } else if (boundingVolume.box) {
+        // TODO: only works for axis aligned boxes
+        const box = boundingVolume.box;
+        const center = new THREE.Vector3(box[0], box[1], box[2]);
+        const w = center.x - box[3];
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/$3dTiles_Provider.js]: Could be worth adding a comment to explain what `box` contain?
It looks like 4 (x, y, z) values, but something like:
```js
// box contains 4 (x, y, z) values, in this order: top-left, bottom-left, top-right, bottom-right
```
helps understanding the code

File: src/Core/Scheduler/Providers/$3dTiles_Provider.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            if (result !== undefined) {
+                let func;
+                const magic = textDecoder.decode(new Uint8Array(result, 0, 4));
+                if (magic[0] === '{') {
+                    result = JSON.parse(textDecoder.decode(new Uint8Array(result)));
+                    const newPrefix = url.slice(0, url.lastIndexOf('/') + 1);
+                    layer.tileIndex.extendTileset(result, metadata.tileId, newPrefix);
+                } else if (magic == 'b3dm') {
+                    func = supportedFormats.b3dm;
+                }
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/$3dTiles_Provider.js]: Add something like:
```js
} else {
Promise.reject(`unsupported magic code ${magic}`);
}
```

File: src/Process/$3dTilesProcessing.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        requestNewTile(context.view, context.scheduler, layer, tileset.root, undefined).then(
+            (tile) => {
+                context.view.scene.add(tile);
+                tile.updateMatrixWorld();
+                layer.root = tile;
+            });
+    });
+}
+
+export function process3dTilesNode(cullingTest, subdivisionTest) {
```
Comment (peppsac) [inline src/Process/$3dTilesProcessing.js]: The whole file looks a lot like `TiledNodeProcessing`.
How about using `TiledNodeProcessing` instead? Adapt it to cover both use-case and maybe rename it `TiledStructureProcessing` ?)

File: examples/3dtiles.html [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            const globe = new itowns.GlobeView(viewerDiv, positionOnGlobe);
+            menuGlobe.view = globe;
+
+            const urlImageryLayers = [];
+            const urlElevationLayers = [];
+
+            itowns.Fetcher.json('layers/JSONLayers/Ortho.json').then(result => globe.addLayer(result));
+
+            itowns.Fetcher.json('layers/JSONLayers/IGN_MNT.json').then(result => globe.addLayer(result));
+            itowns.Fetcher.json('layers/JSONLayers/IGN_MNT_HIGHRES.json').then(result => globe.addLayer(result));
```
Comment (peppsac) [inline examples/3dtiles.html]: Are those needed?
Try to keep examples short, so I'd suggest removing eveything not mandatory for your `3d-tiles` sample:
- elevation layer(s)
- `RealisticLighting` toggle
- `globe.addEventListener(..)`

File: examples/3dtiles.html [hunk 11]
```diff
@@ -0,0 +1,117 @@
+<!DOCTYPE html>
+<!--
+To change this license header, choose License Headers in Project Properties.
+To change this template file, choose Tools | Templates
+and open the template in the editor.
+-->
```
Comment (peppsac) [inline examples/3dtiles.html]: Remove this too

File: examples/3dtiles.html [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            html {height: 100%}
+        body { margin: 0; overflow:hidden; height:100%}
+
+            div {
+                margin : auto auto;
+                width: 100%;
+                padding: 0;
+                height: 100%
+                /*margin-top: 50vh;
+                transform: translateY(-50%);*/
```
Comment (peppsac) [inline examples/3dtiles.html]: And this :)

**Review by peppsac — CHANGES_REQUESTED** [review]
Thanks, this looks great :)

I've left a comments, all small things, except the one on $3dTilesProcessing.js.

File: src/Process/$3dTilesProcessing.js (hunk 9 above)
Comment (Jeremy-Gaillard) [reply src/Process/$3dTilesProcessing.js]: I'd prefer to wait before factorizing. This PR does not cover the full tileset spec, and some of the remaining options that are not yet supported will change the process (I'm mainly thinking about the second, tight bounding box that allows to cull the geometry of the tile without culling the tile itself).

**Comment by Jeremy-Gaillard** [comment]
Pushed modifications asked by @peppsac. If everything is OK, I'll rebase and the PR will be ready to be merged.

File: src/Renderer/ThreeExtended/B3dmLoader.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+const textDecoder = new TextDecoder('utf-8');
+B3dmLoader.prototype.parse = function parse(buffer) {
+    if (!buffer) {
+        throw new Error('No array buffer provided.');
+    }
+
+    const array = new Uint8Array(buffer);
+    const view = new DataView(buffer);
```
Comment (peppsac) [inline src/Renderer/ThreeExtended/B3dmLoader.js]: Using :
```js
// 4 first bytes defined magic type
const array = new Uint8Array(array, 0, 4);
const view = new DataView(buffer, 4);
```
You can remove `byteOffset += 4;` and remove the `subarray` thing.
Actually you could avoid the `array` variable completelty and do:
```js
b3dmHeader.magic = textDecoder.decode(new Uint8Array(buffer, 0, 4));
```
?

**Comment by Jeremy-Gaillard** [comment]
Rebased and ready. The visibility buttons no longer work, but the problem doesn't seem to be specific to 3d-tiles, I guess a recent PR broke it?

**Review by peppsac — APPROVED** [review]
Looks good, merging. Thanks 👍

File: examples/3dtiles.html [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+                itowns.pre3dTilesUpdate(context, layer);
+                return [layer.root];
+            };
+            $3dTilesLayer.update = itowns.process3dTilesNode(
+                itowns.$3dTilesCulling,
+                itowns.$3dTilesSubdivisionControl
+            );
+            $3dTilesLayer.url = 'http://localhost:8003/tilesets/TilesetWithDiscreteLOD/tileset.json';
+            $3dTilesLayer.protocol = '3d-tiles'
+            $3dTilesLayer.overrideMaterials = true;  // custom cesium shaders are not functional
```
Comment (elemoine) [inline examples/3dtiles.html:62]: How is this related to Cesium?

File: examples/3dtiles.html [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            $3dTilesLayer.type = 'geometry';
+            $3dTilesLayer.visible = true;
+
+            itowns.View.prototype.addLayer.call(globe, $3dTilesLayer);
```
Comment (elemoine) [inline examples/3dtiles.html:66]: Why not just do `globe.addLayer($3dTilesLayer)`?

File: examples/3dtiles.html [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+                        n = n[0];
+                    }
+                    n.visible = true;
+                }
+                if (n.length > 0) {
+                    n[0].visible = false;
+                }
+            }
+
+            itowns.View.prototype.addLayer.call(globe, debug3dt, $3dTilesLayer);
```
Comment (elemoine) [inline examples/3dtiles.html:90]: Any reason for adding `$3dTilesLayers` twice?

**Review by elemoine — COMMENTED** [review]
A few post-merge questions, if you like answering them :)

File: examples/3dtiles.html (hunk 14 above)
Comment (Jeremy-Gaillard) [reply examples/3dtiles.html:62]: Cesium uses specific uniforms in its shader to record certain parameters (the sun's position for example). The 3d-tiles samples given by AGI use these uniforms in their shader.

The optimal solution would be to have a compatibility mode where we handle the cesium uniforms in our materials. In the meantime, I just implemented a way to override the 3d-tiles materials.

File: examples/3dtiles.html (hunk 15 above)
Comment (Jeremy-Gaillard) [reply examples/3dtiles.html:66]: The GlobeView class [overrides the addLayer function](https://github.com/iTowns/itowns2/blob/master/src/Core/Prefab/GlobeView.js#L256). Instead of adding a layer to the scene, it attaches the layer to the globe tile layer.

If we want to add a layer that is independent from the tile layer, we have to use this syntax.

File: examples/3dtiles.html (hunk 16 above)
Comment (Jeremy-Gaillard) [reply examples/3dtiles.html:90]: This line doesn't add `$3dTilesLayers` again, it attaches `debug3dt` to `$3dTilesLayers`.

File: examples/3dtiles.html (hunk 15 above)
Comment (elemoine) [reply examples/3dtiles.html:66]: Ok. I find this awkward and hard to understand to example reader.

**Comment by elemoine** [comment]
Thank you for your answers @Jeremy-Gaillard.

File: src/Process/3dTilesProcessing.js [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+
+let preSSE;
+export function pre3dTilesUpdate(context) {
+    // pre-sse
+    const hypotenuse = Math.sqrt(context.camera.width * context.camera.width + context.camera.height * context.camera.height);
+    const radAngle = context.camera.FOV * Math.PI / 180;
+
+     // TODO: not correct -> see new preSSE
+    // const HFOV = 2.0 * Math.atan(Math.tan(radAngle * 0.5) / context.camera.ratio);
+    const HYFOV = 2.0 * Math.atan(Math.tan(radAngle * 0.5) * hypotenuse / context.camera.width);
```
Comment (elemoine) [inline src/Process/3dTilesProcessing.js:68]: From the [three.js doc](https://threejs.org/docs/index.html#api/cameras/PerspectiveCamera) I understand that `context.camera.FOV` (`radAngle`) is the vertical field of view. So I wonder why `HYFOV` is not calculated in terms of `camera.height` instead of `camera.width`:

    const HYFOV = 2.0 * Math.atan(Math.tan(radAngle * 0.5) * hypotenuse / context.camera.height);

?

Also, I wonder why you use the hypotenuse FOV for the SSE calculation. Why not using the vertical FOV as in Cesium? Just curious…

File: src/Process/3dTilesProcessing.js [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+function computeNodeSSE(camera, node) {
+    if (node.boundingVolume.region) {
+        // TODO
+    }
+    if (node.boundingVolume.box) {
+        // TODO: compute proper distance
+        const worldCoordinateCenter = node.boundingVolume.box.getCenter();
+        worldCoordinateCenter.applyMatrix4(node.matrixWorld);
+        const distance = camera.camera3D.position.distanceTo(worldCoordinateCenter);
+        return preSSE * (node.geometricError / distance);
```
Comment (elemoine) [inline src/Process/3dTilesProcessing.js:81]: I also don't understand this. Could you please explain how you got to that expression? The [calculation](https://github.com/AnalyticalGraphicsInc/cesium/blob/38dab936cff23e8f99fb79a3873adfe9400d3bab/Source/Scene/QuadtreePrimitive.js#L688) in Cesium makes more sense to me, but I'd like to understand your reasoning. Thanks.

Comment (Jeremy-Gaillard) [reply src/Process/3dTilesProcessing.js:81]: These functions are transpositions of the terrain tile SSE computations (we'll factorize them in the future). Most of this code was not originally written by me, so I may not be able to answer fully to your questions.

- I think we use the hypotenuse FOV because it is proportional to both the screen height and width
- ~~I remember checking the calculation myself, and it did match with [what I saw elsewhere](http://folk.uio.no/inftt/Div/visualization2.pdf). However, the calculation could be more precise (we compute the SSE for the point at the center of the box instead of the box itself).~~ I checked again, and it seems we're multiplying by `(2.0 * Math.tan(HYFOV * 0.5))` instead of dividing by it (like it is done in Cesium and the presentation I linked). This seems to be a mistake, do you agree @gchoqueux @peppsac?
